src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+    void                 *newpkgcb_data;// newpkgcb data passed in from user
+    cr_XmlParserPkgCb    pkgcb;         // pkgcb passed in from user
+    void                 *pkgcb_data;   // pkgcb data passed in from user
+} cr_CbData;
+
+static int
+call_user_callback_if_package_finished(cr_Package *pkg, cr_CbData *cb_data, GError **err)
+{
+    if ((pkg->loadingflags & CR_PACKAGE_LOADED_PRI) &&
+        (pkg->loadingflags & CR_PACKAGE_LOADED_OTH) &&
+        (pkg->loadingflags & CR_PACKAGE_LOADED_FIL))
+    {
+        g_hash_table_remove(cb_data->in_progress_pkgs, pkg->pkgId);
+
+        // call user packace callback
+        GError *tmp_err = NULL;
+        if (cb_data->pkgcb && cb_data->pkgcb(pkg, cb_data->pkgcb_data, &tmp_err)) {
+            if (tmp_err)
+                g_propagate_prefixed_error(err, tmp_err, "Parsing interrupted: ");
+            else
+                g_set_error(err, ERR_DOMAIN, CRE_CBINTERRUPTED, "Parsing interrupted");
+            return CR_CB_RET_ERR;
+        } else {
+            // If callback return CRE_OK but it simultaneously set
+            // the tmp_err then it's a programming error.
+            assert(tmp_err == NULL);
+        };
+    }
+    return CR_CB_RET_OK;
+}
+
+static int
+newpkgcb_primary(cr_Package **pkg,
+                    G_GNUC_UNUSED const char *pkgId,
+                    G_GNUC_UNUSED const char *name,
+                    G_GNUC_UNUSED const char *arch,
+                    G_GNUC_UNUSED void *cbdata,
+                    GError **err)
+{
+    assert(pkg && *pkg == NULL);
+    assert(!err || *err == NULL);
+
+    *pkg = cr_package_new();
+
+    return CR_CB_RET_OK;
+}
+
+static int
+newpkgcb_general(cr_Package **pkg,
+                    const char *pkgId,
+                    G_GNUC_UNUSED const char *name,
+                    G_GNUC_UNUSED const char *arch,
+                    void *cbdata,
+                    GError **err)
+{
+    assert(pkg && *pkg == NULL);
+    assert(!err || *err == NULL);
+
+    cr_CbData *cb_data = cbdata;
+
+    gpointer pkey = NULL;
+    gpointer pval = NULL;
+    if (g_hash_table_lookup_extended(cb_data->in_progress_pkgs, pkgId, &pkey, &pval)) {
+        // package was already encountred in some other metadata type
+
+        if (pval == NULL) {
+            assert(0);
+            //this pkg was already fully handled
+            //TODO(amatej): maybe error?
+        } else {
+            *pkg = (cr_Package *) pval;
+        }
+    } else {
+        // we are handling never before seen package
+
+        if (cb_data->newpkgcb) {
+            // call user specified new function if provided
+            if (cb_data->newpkgcb(pkg, pkgId, name, arch, cb_data->newpkgcb_data, err)) {
+                return CR_CB_RET_ERR;
+            }
+        } else {
+            *pkg = cr_package_new();
+        }
+        g_hash_table_insert(cb_data->in_progress_pkgs, g_strdup(pkgId), *pkg);
+    }
+
```

Comment:
I would like to better understand the distinction between this `newpkgcb_general` and `primary_pkgcb`.  As far as I can tell, one distinction is that `newpkgcb_general` is used for filelists and other but *not* primary, because primary doesn't provide pkgid as an attribute - so handling that leads to some code duplication.

Is that correct or is it a poor / incomplete understanding?

Would it be valid for this callback to set `pkg->loadingflags |= CR_PACKAGE_FROM_XML;` here?